codec/src/main/java/io/netty/handler/codec/base64/Base64.java

```diff
@@ -168,30 +168,110 @@ private static void encode3to4(
         // significant bytes passed in the array.
         // We have to shift left 24 in order to flush out the 1's that appear
         // when Java treats a value as negative that is cast from a byte to an int.
-        int inBuff =
-                (numSigBytes > 0? src.getByte(srcOffset)     << 24 >>>  8 : 0) |
-                (numSigBytes > 1? src.getByte(srcOffset + 1) << 24 >>> 16 : 0) |
-                (numSigBytes > 2? src.getByte(srcOffset + 2) << 24 >>> 24 : 0);
+        if (src.order() == ByteOrder.BIG_ENDIAN) {
+            final int inBuff;
+            switch (numSigBytes) {
+                case 1:
+                    inBuff = toInt(src.getByte(srcOffset));
+                    break;
+                case 2:
+                    inBuff = toIntBE(src.getShort(srcOffset));
+                    break;
+                default:
+                    inBuff = numSigBytes <= 0 ? 0 : toIntBE(src.getMedium(srcOffset));
+                    break;
+            }
+            encode3to4BigEndian(inBuff, numSigBytes, dest, destOffset, alphabet);
+        } else {
+            final int inBuff;
+            switch (numSigBytes) {
+                case 1:
+                    inBuff = toInt(src.getByte(srcOffset));
+                    break;
+                case 2:
+                    inBuff = toIntLE(src.getShort(srcOffset));
+                    break;
+                default:
+                    inBuff = numSigBytes <= 0 ? 0 : toIntLE(src.getMedium(srcOffset));
+                    break;
+            }
+            encode3to4LittleEndian(inBuff, numSigBytes, dest, destOffset, alphabet);
+        }
+    }
+
+    private static int toInt(byte value) {
+        return (value & 0xff) << 16;
+    }
+
+    private static int toIntBE(short value) {
+        return (value & 0xff00) << 8 | (value & 0xff) << 8;
+    }
+
+    private static int toIntLE(short value) {
+        return (value & 0xff) << 16 | (value & 0xff00);
+    }
+
+    private static int toIntBE(int value) {
+        return (value & 0xff0000) | (value & 0xff00) | (value & 0xff);
+    }
+
+    private static int toIntLE(int value) {
```

Comment:
consider renaming `value` to `mediumValue` or somehow indicate the value only occupies 24 bits.